### PR TITLE
Contracts: connectL1Contracts & connectL2Contracts 

### PR DESCRIPTION
**Pull request by gigamesh**
adds `connectL1Contracts` and `connectL2Contracts` functions to easily connect a provider/signer to any Optimism contracts

### CODE REVIEW

**Comment by snario** [comment]
We shouldn't be adding the artifacts to the version history unnecessarily; we can gitignore them and require the build step.

File: packages/contracts/src/connect-contracts.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  }
+}
+
+export const connectL2Contracts = async (
+  signerOrProvider: Signer | providers.Provider
+): Promise<Object> => {
+  const l2ContractData = getL2ContractData()
+
+  return {
+    eth: new Contract(
```
Comment (gigamesh) [inline packages/contracts/src/connect-contracts.ts:90]: I'm unsure about naming. I think it's probably good to shorten for convenience and remove the "L1" and "L2" from contract names since the user will be connecting to each layer with separate functions?  

But it might add confusion to stray from the .sol names. ex: `OVM_L2CrossDomainMessenger` -> `xDomainMessenger`

File: packages/contracts/test/connect-contracts.spec.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+describe('connectL1Contracts', () => {
+  let user: Signer
+  before(async () => {
+    ;[user] = await ethers.getSigners()
+  })
+
+  it('should connect to all mainnet l1 contracts', async () => {
+    const l1Contracts = await connectL1Contracts(user, 'mainnet')
+    console.log(l1Contracts)
```
Comment (gigamesh) [inline packages/contracts/test/connect-contracts.spec.ts:16]: I'm unable to get any further than this. I can't figure out what is causing this error:

`SyntaxError: Unexpected token x in JSON at position 1`

File: packages/contracts/package.json [hunk 3]
```diff
@@ -16,8 +16,9 @@
   "types": "dist/index",
   "license": "MIT",
   "scripts": {
-    "all": "yarn clean && yarn build && yarn test && yarn lint:fix && yarn lint",
-    "build": "./scripts/build.sh",
+    "all": "yarn build && yarn test && yarn lint:fix && yarn lint",
+    "build": "yarn clean && ./scripts/build.sh && yarn copy-artifacts",
+    "copy-artifacts": "cp -R artifacts-ovm dist/ && cp -R deployments dist/",
```
Comment (snario) [inline packages/contracts/package.json:21]: We can add these to the `files` field of `package.json` I think? This seems like it need not be a user-driven command and instead should be something as part of an automation of some other script.

File: packages/contracts/package.json [hunk 4]
```diff
@@ -16,8 +16,9 @@
   "types": "dist/index",
   "license": "MIT",
   "scripts": {
-    "all": "yarn clean && yarn build && yarn test && yarn lint:fix && yarn lint",
-    "build": "./scripts/build.sh",
+    "all": "yarn build && yarn test && yarn lint:fix && yarn lint",
```
Comment (snario) [inline packages/contracts/package.json:19]: Why did you remove `clean`?

File: packages/contracts/src/connect-contracts.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
+import { Signer, Contract, providers } from 'ethers'
+import { getL1ContractData, getL2ContractData } from './contract-data'
+
+export const connectL1Contracts = async (
+  signerOrProvider: Signer | providers.Provider,
+  network: 'goerli' | 'kovan' | 'mainnet'
+): Promise<Object> => {
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts:7]: We ought to use a more strongly typed object than `Object`.

File: packages/contracts/src/connect-contracts.ts (hunk 1 above)
Comment (snario) [reply packages/contracts/src/connect-contracts.ts:90]: Good questions... would like @smartcontracts's input.

File: packages/contracts/src/connect-contracts.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    erc1820Registry: new Contract(
+      l2ContractData.ERC1820Registry.address,
+      l2ContractData.ERC1820Registry.abi,
+      signerOrProvider
+    ),
+    addressManager: new Contract(
+      l2ContractData.Lib_AddressManager.address,
+      l2ContractData.Lib_AddressManager.abi,
+      signerOrProvider
+    ),
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts:134]: This format is copied throughout the file. Maybe a local helper function like:

```ts
const toEthersContract = ({address, abi}) => new Contract(address, abi, signerOrProvider)
```

File: packages/contracts/src/contract-data.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import * as path from 'path'
+
+export const getL1ContractData = (network: 'goerli' | 'kovan' | 'mainnet') => {
```
Comment (snario) [inline packages/contracts/src/contract-data.ts:3]: Perhaps a more common term is "artifact" ass opposed to "data"?

File: packages/contracts/src/contract-data.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  return {
+    Lib_AddressManager: require(path.resolve(
+      __dirname,
+      `../deployments/${network}-v2/Lib_AddressManager.json`
+    )),
```
Comment (snario) [inline packages/contracts/src/contract-data.ts:8]: May be able to simplify with a helper function here...

```ts
return map(
  ['Lib_AddressManager', 'OVM_CanonicalTransactionChain', ...],
  (contractName) => 
    require(path.resolve(__dirname, '../deployments/${network}-v2/{contractName}.json))
)
```

File: packages/contracts/src/contract-data.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
+import * as path from 'path'
+
+export const getL1ContractData = (network: 'goerli' | 'kovan' | 'mainnet') => {
+  return {
+    Lib_AddressManager: require(path.resolve(
+      __dirname,
+      `../deployments/${network}-v2/Lib_AddressManager.json`
```
Comment (snario) [inline packages/contracts/src/contract-data.ts:7]: Isn't `-v2` a hack? How do we know this will stay the correct folder name? Does it work locally / against an `l1chain`?

File: packages/contracts/tsconfig.json [hunk 10]
```diff
@@ -3,6 +3,6 @@
   "compilerOptions": {
       "resolveJsonModule": true
   },
-  "include": ["./test"],
+  "include": ["./test", "src/**/*"],
```
Comment (snario) [inline packages/contracts/tsconfig.json:6]: Why `src/**/*` instead of `./src`? Curious if there is a difference (I'm never quite sure with this field...)

File: packages/contracts/test/connect-contracts.spec.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+describe('connectL1Contracts', () => {
+  let user: Signer
+  before(async () => {
+    ;[user] = await ethers.getSigners()
+  })
+
+  it('should connect to all mainnet l1 contracts', async () => {
+    const l1Contracts = await connectL1Contracts(user, 'mainnet')
+  })
+})
```
Comment (snario) [inline packages/contracts/test/connect-contracts.spec.ts:14]: Need to add a test for `connectL2Contracts`

File: packages/contracts/src/contract-data.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        `../artifacts-ovm/contracts/optimistic-ethereum/OVM/predeploys/OVM_L2ToL1MessagePasser.sol/OVM_L2ToL1MessagePasser.json`
+      )),
+      address: '0x4200000000000000000000000000000000000000',
+    },
+    OVM_L1MessageSender: {
+      abi: require(path.resolve(
+        __dirname,
+        `../artifacts-ovm/contracts/optimistic-ethereum/OVM/predeploys/OVM_L1MessageSender.sol/OVM_L1MessageSender.json`
+      )),
+      address: '0x4200000000000000000000000000000000000001',
```
Comment (snario) [inline packages/contracts/src/contract-data.ts:92]: These ought to be accessible in some constants file somewhere, no? i.e., some function `f` such that `f('OVM_L1MessageSender')` gets `0x4200000000000000000000000000000000000001`? It's a system wide constant

**Review by snario — CHANGES_REQUESTED** [review]
Requested some changes.

This is also making me realize that a logical next step (in a future PR? @gigamesh curious your thoughts here) would be to update the local development environment to rely on this function as opposed to services independently listening for the state dump. Instead, the `contracts` repo could be the thing pulling `addresses.json` from the `deployer` service and then other services would rely on `@eth-optimism/contracts` for retrieving these addresses.

File: packages/contracts/test/connect-contracts.spec.ts (hunk 2 above)
Comment (snario) [reply packages/contracts/test/connect-contracts.spec.ts]: Did you resolve this?

File: packages/contracts/package.json (hunk 4 above)
Comment (gigamesh) [reply packages/contracts/package.json:19]: @snario I figure if its in the `yarn build` script, its redundant to have it in `yarn all`, but happy to revert if you'd like. lmk

**Comment by gigamesh** [comment]
> Requested some changes.
> 
> This is also making me realize that a logical next step (in a future PR? @gigamesh curious your thoughts here) would be to update the local development environment to rely on this function as opposed to services independently listening for the state dump. Instead, the `contracts` repo could be the thing pulling `addresses.json` from the `deployer` service and then other services would rely on `@eth-optimism/contracts` for retrieving these addresses.

sounds good to me!

File: packages/contracts/package.json (hunk 3 above)
Comment (gigamesh) [reply packages/contracts/package.json:21]: ahh yea - I just needed to add `deployments` to the `files` list. thanks

File: packages/contracts/test/connect-contracts.spec.ts (hunk 2 above)
Comment (gigamesh) [reply packages/contracts/test/connect-contracts.spec.ts]: yep Kelvin found the problem

File: packages/contracts/src/contract-data.ts (hunk 9 above)
Comment (gigamesh) [reply packages/contracts/src/contract-data.ts:7]: fixed by removing the old versions from the directory

File: packages/contracts/tsconfig.json (hunk 10 above)
Comment (gigamesh) [reply packages/contracts/tsconfig.json:6]: not sure but that's how it's defined in the example: https://www.typescriptlang.org/tsconfig#include

File: packages/contracts/src/contract-data.ts (hunk 12 above)
Comment (gigamesh) [reply packages/contracts/src/contract-data.ts]: yeah good call - not a contract but grabbed the object from predeploys.ts

File: packages/contracts/test/connect-contracts.spec.ts (hunk 11 above)
Comment (gigamesh) [reply packages/contracts/test/connect-contracts.spec.ts:54]: Not sure if this is enough test coverage. I could also check the addresses are correct, but that would require reaching outside of the test directory (seems like it could cause problems)

File: packages/contracts/src/connect-contracts.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  }
+}
+
+export const connectL2Contracts = async (
+  signerOrProvider: Signer | providers.Provider
+): Promise<{ [key: string]: Contract }> => {
+  const l2ContractData = await getL2ContractData()
+
+  const toEthersContract = (address, abi) =>
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts:78]: You can change this to `({address, abi)}` and then pass in `l2ContractData[contractName]` below

File: packages/contracts/package.json [hunk 14]
```diff
@@ -8,6 +8,7 @@
     "dist/types-ovm/*.ts",
     "artifacts/**/*.json",
     "artifacts-ovm/**/*.json",
+    "deployments/**/*.json",
```
Comment (snario) [inline packages/contracts/package.json:11]: Why are you adding this?

**Comment by gakonst** [comment]
@gigamesh there's also some linting errors which should be fixed!

File: packages/contracts/package.json (hunk 14 above)
Comment (gigamesh) [reply packages/contracts/package.json:11]: this contains all the L1 artifacts & addresses so it needs to be in the published package

File: packages/contracts/src/connect-contracts.ts (hunk 13 above)
Comment (gigamesh) [reply packages/contracts/src/connect-contracts.ts]: Not sure I understood your suggestion but I made another change to make the file more DRY

**Review by snario — APPROVED** [review]
I want to make one last request on something I noticed when trying this out locally which is handling poorly formatted input. For example,

```node
> const { connectL1Contracts } = require('@eth-optimism/contracts')
> connectL1Contracts('mainnet')
Error: Cannot find module '/Users/liam/code/optimism/optimism/packages/contracts/deployments/undefined/Lib_AddressManager.json'
Require stack:
- /Users/liam/code/optimism/optimism/packages/contracts/dist/contract-data.js
- /Users/liam/code/optimism/optimism/packages/contracts/dist/connect-contracts.js
- /Users/liam/code/optimism/optimism/packages/contracts/dist/index.js
- <repl>
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:940:15)
    at Function.Module._load (node:internal/modules/cjs/loader:773:27)
    at Module.require (node:internal/modules/cjs/loader:1012:19)
    at require (node:internal/modules/cjs/helpers:93:18)
    at /Users/liam/code/optimism/optimism/packages/contracts/src/contract-data.ts:26:36
    at Array.reduce (<anonymous>)
    at Object.getL1ContractData (/Users/liam/code/optimism/optimism/packages/contracts/src/contract-data.ts:21:24)
    at Object.connectL1Contracts (/Users/liam/code/optimism/optimism/packages/contracts/src/connect-contracts.ts:8:26)
    at REPL19:1:3
    at sigintHandlersWrap (node:vm:271:12)
    at Script.runInThisContext (node:vm:129:14)
    at REPLServer.defaultEval (node:repl:524:29)
    at bound (node:domain:416:15)
    at REPLServer.runBound [as eval] (node:domain:427:12)
    at REPLServer.onLine (node:repl:846:10)
    at REPLServer.emit (node:events:377:35) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/liam/code/optimism/optimism/packages/contracts/dist/contract-data.js',
    '/Users/liam/code/optimism/optimism/packages/contracts/dist/connect-contracts.js',
    '/Users/liam/code/optimism/optimism/packages/contracts/dist/index.js',
    '<repl>'
  ]
```

Otherwise looks good!

Some bonus tasks:
- Check if `providerOrSigner` has a `chainId` which matches `network`
- Ensure we can provide a custom network name / network (e.g., for local development)
   - Shouldn't we pass in a `chainId` not a `network` btw?

File: packages/contracts/src/connect-contracts.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
+import { Signer, Contract, providers } from 'ethers'
+import { getL1ContractData, getL2ContractData } from './contract-data'
+
+export const connectL1Contracts = async (
+  signerOrProvider: Signer | providers.Provider,
+  network: 'goerli' | 'kovan' | 'mainnet'
+): Promise<{ [key: string]: Contract }> => {
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts:7]: We should be typing this object e.g., https://github.com/graphprotocol/common-ts/blob/master/packages/common-ts/src/contracts/index.ts#L34

**Comment by gigamesh** [comment]
This still has some problems. I think I need a pair program session with someone who has some experience with building npm packages.

File: packages/contracts/deployments/index.d.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    OVM_L1ETHGateway: any;
+    OVM_L1MultiMessageRelayer: any;
+    OVM_SafetyChecker: any;
+    OVM_StateCommitmentChain: any;
+    OVM_StateManagerFactory: any;
+    OVM_StateTransitionerFactory: any;
+    Proxy__OVM_L1CrossDomainMessenger: any;
+    Proxy__OVM_L1ETHGateway: any;
+    mockOVM_BondManager: any;
+};
```
Comment (snario) [inline packages/contracts/deployments/index.d.ts]: Should this be gitignoed?

File: packages/contracts/src/connect-contracts.ts [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+//   ovmETH: typeof OVM_ETH.abi
+//   xDomainMessenger: typeof OVM_L2CrossDomainMessenger.abi
+//   l1MessageSender: typeof OVM_L1MessageSender.abi
+//   l2MessagePasser: typeof OVM_L2ToL1MessagePasser.abi
+//   deployerWhiteList: typeof OVM_DeployerWhitelist.abi
+//   sequencerEntryPoint: typeof OVM_SequencerEntrypoint.abi
+//   ecdsaContract: typeof OVM_ECDSAContractAccount.abi
+//   l2AddressManager: typeof Lib_L2AddressManager.abi
+//   erc1820Registry: typeof ERC1820Registry.abi
+// }
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts]: Why is this commented out?

File: packages/contracts/artifacts-ovm/@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol/OwnableUpgradeable.dbg.json [hunk 18]
```diff
@@ -0,0 +1,4 @@
+{
+  "_format": "hh-sol-dbg-1",
+  "buildInfo": "../../../../build-info/4853042e4eb9d1f9813cc96d52678088.json"
+}
```
Comment (snario) [inline packages/contracts/artifacts-ovm/@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol/OwnableUpgradeable.dbg.json]: Shouldn't this be gitignored?

File: packages/contracts/deployments/index.d.ts (hunk 16 above)
Comment (gigamesh) [reply packages/contracts/deployments/index.d.ts]: this was deleted in a subsequent commit

File: packages/contracts/src/connect-contracts.ts (hunk 17 above)
Comment (gigamesh) [reply packages/contracts/src/connect-contracts.ts]: outdated - I think you're viewing an older commit

**Comment by smartcontracts** [comment]
I also noticed that you removed the various artifacts files. Could we do that in a separate PR? It creates a huge diff and makes it difficult to review the code that you added.

File: packages/contracts/package.json [hunk 19]
```diff
@@ -47,13 +49,14 @@
   "dependencies": {
     "@eth-optimism/core-utils": "^0.4.1",
     "@ethersproject/abstract-provider": "^5.0.8",
+    "@ethersproject/abstract-signer": "^5.1.0",
```
Comment (smartcontracts) [inline packages/contracts/package.json:52]: You imported `abstract-signer` here but it's unused

File: yarn.lock [hunk 20]
```diff
@@ -2,34 +2,908 @@
 # yarn lockfile v1
 
 
-"@babel/code-frame@^7.0.0":
+"@babel/code-frame@^7.0.0", "@babel/code-frame@^7.12.13":
```
[the remark this answers is not in the document]
Comment (smartcontracts) [reply yarn.lock:5]: I see, you added `@ethersproject/abstract-signer` but it seems to be unused. Requesting that we remove that dependency if unused and remove the changes to `yarn.lock` if not strictly necessary (`yarn.lock` is hard to review).

File: packages/contracts/src/contract-data.ts [hunk 21]
```diff
@@ -0,0 +1,174 @@
+import * as path from 'path'
+import { predeploys as l2Addresses } from './predeploys'
+import { Network } from './connect-contracts'
+
+const Mainnet__Lib_AddressManager = require('../deployments/mainnet/Lib_AddressManager.json')
```
[the remark this answers is not in the document]
Comment (gigamesh) [reply packages/contracts/src/contract-data.ts:5]: There is a lot of backstory with this file. Liam requested the same (do the imports with a helper function) but after a ton of struggling, we realized we can't use Node.js stuff (`path`, `fs`, etc) because this is for the browser.

File: packages/contracts/package.json [hunk 22]
```diff
@@ -16,8 +17,8 @@
   "types": "dist/index",
   "license": "MIT",
   "scripts": {
-    "all": "yarn clean && yarn build && yarn test && yarn lint:fix && yarn lint",
-    "build": "./scripts/build.sh",
+    "all": "yarn build && yarn test && yarn lint:fix && yarn lint",
+    "build": "yarn clean && ./scripts/build.sh",
```
[the remark this answers is not in the document]
Comment (gigamesh) [reply packages/contracts/package.json:21]: the only place `yarn clean` was being run was right before `yarn build` in the `all` script so I figured we can just move it directly into build (I need clean to be run before build for the purposes of this PR)

File: packages/contracts/package.json [hunk 23]
```diff
@@ -30,11 +31,12 @@
     "test:contracts": "hardhat test --show-stack-traces",
     "test:gas": "hardhat test \"test/contracts/OVM/execution/OVM_StateManager.gas-spec.ts\" --no-compile --show-stack-traces",
     "test:coverage": "NODE_OPTIONS=--max_old_space_size=8192 hardhat coverage",
+    "test:connect-contracts": "yarn run test:contracts test/connect-contracts.spec.ts",
     "lint": "yarn lint:fix && yarn lint:check",
+    "lint:typescript": "tslint --format stylish --project .",
```
[the remark this answers is not in the document]
Comment (gigamesh) [reply packages/contracts/package.json:36]: accidental

File: yarn.lock (hunk 20 above)
[the remark this answers is not in the document]
Comment (gigamesh) [reply yarn.lock:5]: import { Signer } from '@ethersproject/abstract-signer' is in `hardhat-deploy-ethers.ts` and node was complaining that it was missing

I'm honestly not sure re: babel/code-frame ...I have a very weak understanding of the yarn.lock file and usually just delete it & reinstall when there are problems. lmk if I should be doing something else

**Comment by gigamesh** [comment]
> I also noticed that you removed the various artifacts files. Could we do that in a separate PR? It creates a huge diff and makes it difficult to review the code that you added.

https://github.com/ethereum-optimism/optimism/pull/827

File: packages/contracts/package.json [hunk 24]
```diff
@@ -30,11 +31,12 @@
     "test:contracts": "hardhat test --show-stack-traces",
     "test:gas": "hardhat test \"test/contracts/OVM/execution/OVM_StateManager.gas-spec.ts\" --no-compile --show-stack-traces",
     "test:coverage": "NODE_OPTIONS=--max_old_space_size=8192 hardhat coverage",
+    "test:connect-contracts": "yarn run test:contracts test/connect-contracts.spec.ts",
```
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/package.json:34]: Hmmm I see. I think we should remove this -- during development it should be easy enough to `yarn test ./test/connect-contracts.spec.ts`

**Comment by smartcontracts** [comment]
@gigamesh could you rebase on top of master? That would remove the large diff.

File: packages/contracts/src/contract-data.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/src/contract-data.ts:5]: One option for a helper function here is something along the lines of:

```ts
const loadContractData = (network: string, name: string) => {
  return require(`../deployments/${network}/${name}.json`)
}
```

Then you could update `getL1ContractData` as follows:

```
export const getL1ContractData = (network: Network) => {
  return {
    Lib_AddressManager: loadContractData(network, 'Lib_AddressManager'),
    ...
  }
}
```

[the remark this answers is not in the document]
Comment (gigamesh) [reply packages/contracts/src/contract-data.ts:5]: The problem is all the json needs to be imported when the typescript gets compiled, so I don't think dynamic imports are possible. The browser can't load json into javascript during runtime AFAIK.

[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/src/contract-data.ts:5]: Hmmmm I see, give me a sec to think about this...

File: packages/contracts/package.json (hunk 22 above)
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/package.json:21]: I think for now we should revert this change. If you feel strongly we can make another issue for it and tackle it there. Just want to keep this PR tightly scoped and avoid changes that could potentially cause issues for others (not sure if it actually will, just being safe).

File: packages/contracts/src/connect-contracts.ts [hunk 25]
```diff
@@ -0,0 +1,122 @@
+import { Signer, Contract, providers, ethers } from 'ethers'
+import { Provider } from '@ethersproject/abstract-provider'
+import { getL1ContractData, getL2ContractData } from './contract-data'
+
+export type Network = 'goerli' | 'kovan' | 'mainnet'
+interface L1Contracts {
+  addressManager: Contract
```
Comment (smartcontracts) [inline packages/contracts/src/connect-contracts.ts:7]: I think it might be better to use the full contract names (the upper case ones), e.g., `addressManager` => `Lib_AddressManager` or `AddressManager`, since we're already using that format everywhere else.

File: packages/contracts/src/contract-data.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/src/contract-data.ts:5]: OK for now this is fine, but I'll just request that we add a comment explaining why the code looks the way it looks (or others might have the same reaction that I did).

[the remark this answers is not in the document]
Comment (smartcontracts) [reply packages/contracts/src/contract-data.ts:5]: In the future I think we should consider dynamically generating this file to reduce the amount of code floating around.

File: yarn.lock [hunk 26]
```diff
@@ -9,24 +9,24 @@
   dependencies:
     "@babel/highlight" "^7.12.13"
 
-"@babel/helper-validator-identifier@^7.12.11":
```
Comment (smartcontracts) [inline yarn.lock:12]: Since you got rid of the `@ethersproject/abstract-signer` import we should be able to get rid of this diff.

File: packages/contracts/src/connect-contracts.ts [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+  messagePasser: Contract
+  messageSender: Contract
+  deployerWhiteList: Contract
+  ecdsaContractAccount: Contract
+  sequencerEntrypoint: Contract
+  erc1820Registry: Contract
+  addressManager: Contract
+}
+
+const checkSignerType = (signerOrProvider: Signer | Provider) => {
```
Comment (smartcontracts) [inline packages/contracts/src/connect-contracts.ts:35]: Requesting that you add some comments to these functions.

File: packages/contracts/src/connect-contracts.ts [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+  if (!signerOrProvider) throw Error('signerOrProvider argument is undefined')
```
Comment (smartcontracts) [inline packages/contracts/src/connect-contracts.ts:36]: We're trying to enforce curly brace syntax for `if` statements. I should make a ticket to add a check for curly braces to our linter.

File: packages/contracts/src/connect-contracts.ts (hunk 25 above)
Comment (gigamesh) [reply packages/contracts/src/connect-contracts.ts:7]: I actually have a pretty strong opinion about this for a few reasons:
- in javascript, capital case usually means something is a class (and/or component, like a react component), but they get returned as objects 
- the synthetix team (I think Justin) was ribbing us at one point about our naming scheme because we're mixing underscores and camel case
- the contract names are longer, so not very dev-ex friendly

...I figure if this is for external use, we should use names that make more sense in a javascript context (rather than solidity). that said, I'm fine using the contract names if everyone else prefers it.

Comment (smartcontracts) [reply packages/contracts/src/connect-contracts.ts:7]: Sure, we can leave it at this for now. Worst case we can always use both.

**Review by smartcontracts — APPROVED** [review]

**Comment by smartcontracts** [comment]
Converting to draft since it seems we're now working on this in a few different pieces

**Comment by gigamesh** [comment]
> Converting to draft since it seems we're now working on this in a few different pieces

@smartcontracts Any ideas why this is still failing?

**Comment by maurelian** [comment]
Is this a duplicate of #858 that we can close?

**Comment by snario** [comment]
No, this is larger in scope. cc @gigamesh

**Comment by gigamesh** [comment]
> Is this a duplicate of #858 that we can close?

#858 is an attempt to merge some of the work from this PR and narrow down why the CI is failing. I haven't been able to figure out what the problem is.

**Comment by snario** [comment]
There is an error being thrown in `deployer`

File: packages/contracts/src/connect-contracts.ts [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+  xDomainMessenger: Contract
+  ethGateway: Contract
+  multiMessageRelayer: Contract
+  safetyChecker: Contract
+  stateCommitmentChain: Contract
+  stateManagerFactory: Contract
+  stateTransitionerFactory: Contract
+  xDomainMessengerProxy: Contract
+  l1EthGatewayProxy: Contract
+  mockBondManger: Contract
```
Comment (snario) [inline packages/contracts/src/connect-contracts.ts:20]: Is this really the list we want to expose here? We may not need to expose e.g., `mockBondManager` (which has a typo btw) and the factories perhaps?

File: packages/contracts/src/connect-contracts.ts [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+    )
+    network = 'mainnet'
+  }
+  if (network !== 'mainnet' && network !== 'kovan' && network !== 'goerli') {
+    throw Error('network argument is the wrong type')
+  }
+
+  const l1ContractData = getL1ContractData(network)
+
+  const toEthersContract = (data) =>
```
Comment (tynes) [inline packages/contracts/src/connect-contracts.ts:75]: Nit: this function is defined twice

File: packages/contracts/src/connect-contracts.ts [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+export const connectL1Contracts = async (
+  signerOrProvider,
+  network?: Network
+): Promise<L1Contracts> => {
+  checkSignerType(signerOrProvider)
+
+  if (!network) {
+    console.warn(
+      'network argument not provided to connectL1Contracts. Defaulting to mainnet.'
+    )
```
Comment (tynes) [inline packages/contracts/src/connect-contracts.ts:66]: I'm not sure how I feel about this warning - its a good idea for preventing bugs but will end up in writing to stdout when the application may not be expecting a library to write to stdout. Generally ok for browser usage though

**Comment by tynes** [comment]
This needs a changeset still

File: packages/contracts/src/connect-contracts.ts (hunk 29 above)
Comment (smartcontracts) [reply packages/contracts/src/connect-contracts.ts:20]: I would say we should expose the following:

```
addressManager
canonicalTransactionChain
stateCommitmentChain
executionManager
fraudVerifier
l1EthGatewayProxy
l1CrossDomainMessengerProxy
mulitMessageRelayer
bondManager
```

Comment (smartcontracts) [reply packages/contracts/src/connect-contracts.ts:20]: We definitely should NOT expose the non-proxied contracts since we've been working hard to keep those hidden as to avoid situations where people interact with the implementations instead of the proxies...

File: packages/contracts/src/connect-contracts.ts (hunk 31 above)
Comment (smartcontracts) [reply packages/contracts/src/connect-contracts.ts:66]: Yeah IMO `network` should be a required parameter if we're considering having a warning like this.

File: packages/contracts/src/connect-contracts.ts [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+    network = 'mainnet'
+  }
+  if (network !== 'mainnet' && network !== 'kovan' && network !== 'goerli') {
```
Comment (smartcontracts) [inline packages/contracts/src/connect-contracts.ts:69]: Alternatively:

```
if (!(network in ['mainnet', 'kovan', 'goerli'])) { ... }
```

**Review by smartcontracts — CHANGES_REQUESTED** [review]
Requesting a change to the contracts that we expose. Otherwise looking good.

File: packages/contracts/src/connect-contracts.ts (hunk 30 above)
Comment (gigamesh) [reply packages/contracts/src/connect-contracts.ts:75]: It's technically 2 definitions. The signerOrProvider comes from outside its scope. 

I don't have a strong preference but Liam asked for several changes to minimize code. If the signerOrProvider is required as an argument, it results in a lot of repetition and ends up kind of becoming a pointless helper.

**Review by smartcontracts — APPROVED** [review]
Looks good!

File: packages/contracts/src/connect-contracts.ts [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+ * @param network string denoting network
+ * @returns l1 contracts connected to signer/provider
+ */
+export const connectL1Contracts = async (
+  signerOrProvider: Signer | Provider,
+  network: Network
+): Promise<L1Contracts> => {
+  checkSignerType(signerOrProvider)
+
+  console.log(network)
```
Comment (tynes) [inline packages/contracts/src/connect-contracts.ts:59]: dead `console.log`

File: packages/contracts/src/connect-contracts.ts (hunk 30 above)
Comment (tynes) [reply packages/contracts/src/connect-contracts.ts:66]: Does making `signerOrProvider` be a optional arg help? ie `signerOrProvider?`

File: packages/contracts/test/connect-contracts.spec.ts [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+      expect(l1Contracts[name]).to.be.an.instanceOf(Contract)
+    })
+  }
+
+  for (const name of l2ContractNames) {
+    it(`connectL2Contracts should return a contract assigned to a field named "${name}"`, async () => {
+      const l2Contracts = await connectL2Contracts(user)
+      expect(l2Contracts[name]).to.be.an.instanceOf(Contract)
+    })
+  }
```
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts/test/connect-contracts.spec.ts:60]: cc @gigamesh

File: packages/contracts/src/connect-contracts.ts (hunk 33 above)
Comment (gigamesh) [reply packages/contracts/src/connect-contracts.ts]: oops - removed

File: .changeset/fair-icons-argue.md [hunk 35]
```diff
@@ -0,0 +1,5 @@
+---
+'@eth-optimism/contracts': minor
```
Comment (snario) [inline .changeset/fair-icons-argue.md:2]: This should be a `patch`

**Review by snario — CHANGES_REQUESTED** [review]

**Review by snario — DISMISSED** [review]

File: packages/contracts/src/contract-data.ts [hunk 36]
```diff
@@ -0,0 +1,180 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
```
Comment (gigamesh) [inline packages/contracts/src/contract-data.ts:1]: I tried switching the require statements to `import` but I couldn't get tslint to stop complaining. It said to add a list of `includes` to the config, which I tried, but still didn't work.

**Review by snario — APPROVED** [review]

**Comment by gigamesh** [comment]
@smartcontracts do you mind giving me an approval on this? :)
